File: project_management_service/backend/src/models/projects.rs
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// Required Notice: Copyright (c) 2025 AI Chat Team

use serde::{Deserialize, Serialize};

use super::DbStatus;

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ProjectStatus {
    #[default]
    Active,
    Archived,
}

impl DbStatus for ProjectStatus {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Active => "active",
            Self::Archived => "archived",
        }
    }

    fn from_db(value: &str) -> Self {
        match value.trim() {
            "archived" => Self::Archived,
            _ => Self::Active,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ProjectSourceType {
    Local,
    LocalConnector,
    #[default]
    Cloud,
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ProjectExecutionPlane {
    LocalConnector,
    #[default]
    Cloud,
}

impl DbStatus for ProjectExecutionPlane {
    fn as_str(&self) -> &'static str {
        match self {
            Self::LocalConnector => "local_connector",
            Self::Cloud => "cloud",
        }
    }

    fn from_db(value: &str) -> Self {
        match value.trim() {
            "local_connector" => Self::LocalConnector,
            _ => Self::Cloud,
        }
    }
}

impl ProjectSourceType {
    pub fn execution_plane(self) -> ProjectExecutionPlane {
        match self {
            Self::Cloud => ProjectExecutionPlane::Cloud,
            Self::Local | Self::LocalConnector => ProjectExecutionPlane::LocalConnector,
        }
    }
}

impl DbStatus for ProjectSourceType {
    fn as_str(&self) -> &'static str {
        match self {
            Self::Local => "local",
            Self::LocalConnector => "local_connector",
            Self::Cloud => "cloud",
        }
    }

    fn from_db(value: &str) -> Self {
        match value.trim() {
            "local" => Self::Local,
            "local_connector" => Self::LocalConnector,
            _ => Self::Cloud,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum CloudImportSource {
    #[default]
    None,
    Empty,
    Git,
    Zip,
}

impl DbStatus for CloudImportSource {
    fn as_str(&self) -> &'static str {
        match self {
            Self::None => "none",
            Self::Empty => "empty",
            Self::Git => "git",
            Self::Zip => "zip",
        }
    }

    fn from_db(value: &str) -> Self {
        match value.trim() {
            "empty" => Self::Empty,
            "git" => Self::Git,
            "zip" => Self::Zip,
            _ => Self::None,
        }
    }
}

#[derive(Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "snake_case")]
#[derive(Default)]
pub enum ProjectImportStatus {
    #[default]
    None,
    Pending,
    Importing,
    Ready,
    Failed,
}

impl DbStatus for ProjectImportStatus {
    fn as_str(&self) -> &'static str {
        match self {
            Self::None => "none",
            Self::Pending => "pending",
            Self::Importing => "importing",
            Self::Ready => "ready",
            Self::Failed => "failed",
        }
    }

    fn from_db(value: &str) -> Self {
        match value.trim() {
            "pending" => Self::Pending,
            "importing" => Self::Importing,
            "ready" => Self::Ready,
            "failed" => Self::Failed,
            _ => Self::None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectRecord {
    pub id: String,
    #[serde(default)]
    pub creator_user_id: Option<String>,
    #[serde(default)]
    pub creator_username: Option<String>,
    #[serde(default)]
    pub creator_display_name: Option<String>,
    pub owner_user_id: Option<String>,
    pub owner_username: Option<String>,
    pub owner_display_name: Option<String>,
    pub name: String,
    pub root_path: Option<String>,
    pub git_url: Option<String>,
    #[serde(default)]
    pub source_type: ProjectSourceType,
    #[serde(default)]
    pub execution_plane: ProjectExecutionPlane,
    #[serde(default)]
    pub cloud_import_source: CloudImportSource,
    #[serde(default)]
    pub import_status: ProjectImportStatus,
    #[serde(default)]
    pub source_git_url: Option<String>,
    #[serde(default)]
    pub harness_space_identifier: Option<String>,
    #[serde(default)]
    pub harness_repo_identifier: Option<String>,
    #[serde(default)]
    pub harness_repo_path: Option<String>,
    #[serde(default)]
    pub harness_git_url: Option<String>,
    #[serde(default)]
    pub harness_git_ssh_url: Option<String>,
    #[serde(default)]
    pub harness_default_branch: Option<String>,
    #[serde(default)]
    pub harness_provision_status: Option<String>,
    #[serde(default)]
    pub harness_provision_error: Option<String>,
    #[serde(default)]
    pub harness_provisioned_at: Option<String>,
    #[serde(default)]
    pub import_error: Option<String>,
    #[serde(default)]
    pub import_started_at: Option<String>,
    #[serde(default)]
    pub import_finished_at: Option<String>,
    pub description: Option<String>,
    pub status: ProjectStatus,
    pub created_at: String,
    pub updated_at: String,
    pub archived_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct CreateProjectRequest {
    pub name: String,
    pub root_path: Option<String>,
    pub git_url: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub sandbox_enabled: Option<bool>,
    #[serde(default)]
    pub source_type: Option<ProjectSourceType>,
    #[serde(default)]
    pub cloud_import_source: Option<CloudImportSource>,
    #[serde(default)]
    pub import_status: Option<ProjectImportStatus>,
    #[serde(default)]
    pub source_git_url: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImportProjectRequest {
    pub id: String,
    pub owner_user_id: Option<String>,
    pub owner_username: Option<String>,
    pub owner_display_name: Option<String>,
    pub name: String,
    pub root_path: Option<String>,
    pub git_url: Option<String>,
    pub description: Option<String>,
    #[serde(default)]
    pub sandbox_enabled: Option<bool>,
    #[serde(default)]
    pub source_type: Option<ProjectSourceType>,
    #[serde(default)]
    pub cloud_import_source: Option<CloudImportSource>,
    #[serde(default)]
    pub import_status: Option<ProjectImportStatus>,
    #[serde(default)]
    pub source_git_url: Option<String>,
    #[serde(default)]
    pub harness_space_identifier: Option<String>,
    #[serde(default)]
    pub harness_repo_identifier: Option<String>,
    #[serde(default)]
    pub harness_repo_path: Option<String>,
    #[serde(default)]
    pub harness_git_url: Option<String>,
    #[serde(default)]
    pub harness_git_ssh_url: Option<String>,
    #[serde(default)]
    pub harness_default_branch: Option<String>,
    #[serde(default)]
    pub harness_provision_status: Option<String>,
    #[serde(default)]
    pub harness_provision_error: Option<String>,
    #[serde(default)]
    pub harness_provisioned_at: Option<String>,
    #[serde(default)]
    pub import_error: Option<String>,
    #[serde(default)]
    pub import_started_at: Option<String>,
    #[serde(default)]
    pub import_finished_at: Option<String>,
    pub status: Option<ProjectStatus>,
    pub created_at: Option<String>,
    pub updated_at: Option<String>,
    pub archived_at: Option<String>,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UpdateProjectRequest {
    pub name: Option<String>,
    pub root_path: Option<String>,
    pub git_url: Option<String>,
    pub description: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ProjectProfileRecord {
    pub project_id: String,
    #[serde(default)]
    pub creator_user_id: Option<String>,
    #[serde(default)]
    pub creator_username: Option<String>,
    #[serde(default)]
    pub creator_display_name: Option<String>,
    #[serde(default)]
    pub owner_user_id: Option<String>,
    #[serde(default)]
    pub owner_username: Option<String>,
    #[serde(default)]
    pub owner_display_name: Option<String>,
    pub background: Option<String>,
    pub introduction: Option<String>,
    pub created_at: String,
    pub updated_at: String,
}

#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct UpsertProjectProfileRequest {
    pub background: Option<String>,
    pub introduction: Option<String>,
}

File: chatos/frontend/src/lib/domain/projects.ts
// SPDX-License-Identifier: PolyForm-Noncommercial-1.0.0
// Required Notice: Copyright (c) 2025 AI Chat Team

import type { Project } from '../../types';
import type { ProjectResponse } from '../api/client/types';
import {
  asRecord,
  normalizeDate,
  readValue,
} from './normalizerUtils';

export const normalizeProject = (raw: ProjectResponse | unknown): Project => {
  const record = asRecord(raw);
  const createdAtSource = readValue(record, 'created_at') ?? readValue(record, 'createdAt') ?? Date.now();
  const updatedAtSource = readValue(record, 'updated_at')
    ?? readValue(record, 'updatedAt')
    ?? createdAtSource;

  return {
    id: (readValue(record, 'id') ?? '') as Project['id'],
    name: (readValue(record, 'name') ?? '') as Project['name'],
    rootPath: (readValue(record, 'root_path') ?? readValue(record, 'rootPath') ?? '') as Project['rootPath'],
    displayRootPath: (readValue(record, 'display_root_path') ?? readValue(record, 'displayRootPath') ?? null) as Project['displayRootPath'],
    gitUrl: (readValue(record, 'git_url') ?? readValue(record, 'gitUrl') ?? null) as Project['gitUrl'],
    sourceType: (readValue(record, 'source_type') ?? readValue(record, 'sourceType') ?? 'cloud') as Project['sourceType'],
    executionPlane: (readValue(record, 'execution_plane') ?? readValue(record, 'executionPlane') ?? null) as Project['executionPlane'],
    cloudImportSource: (readValue(record, 'cloud_import_source') ?? readValue(record, 'cloudImportSource') ?? null) as Project['cloudImportSource'],
    importStatus: (readValue(record, 'import_status') ?? readValue(record, 'importStatus') ?? null) as Project['importStatus'],
    sourceGitUrl: (readValue(record, 'source_git_url') ?? readValue(record, 'sourceGitUrl') ?? null) as Project['sourceGitUrl'],
    harnessSpaceIdentifier: (readValue(record, 'harness_space_identifier') ?? readValue(record, 'harnessSpaceIdentifier') ?? null) as Project['harnessSpaceIdentifier'],
    harnessRepoIdentifier: (readValue(record, 'harness_repo_identifier') ?? readValue(record, 'harnessRepoIdentifier') ?? null) as Project['harnessRepoIdentifier'],
    harnessRepoPath: (readValue(record, 'harness_repo_path') ?? readValue(record, 'harnessRepoPath') ?? null) as Project['harnessRepoPath'],
    harnessGitUrl: (readValue(record, 'harness_git_url') ?? readValue(record, 'harnessGitUrl') ?? null) as Project['harnessGitUrl'],
    harnessGitSshUrl: (readValue(record, 'harness_git_ssh_url') ?? readValue(record, 'harnessGitSshUrl') ?? null) as Project['harnessGitSshUrl'],
    importError: (readValue(record, 'import_error') ?? readValue(record, 'importError') ?? null) as Project['importError'],
    importStartedAt: (readValue(record, 'import_started_at') ?? readValue(record, 'importStartedAt') ?? null) as Project['importStartedAt'],
    importFinishedAt: (readValue(record, 'import_finished_at') ?? readValue(record, 'importFinishedAt') ?? null) as Project['importFinishedAt'],
    description: (readValue(record, 'description') ?? null) as Project['description'],
    userId: (readValue(record, 'user_id') ?? readValue(record, 'userId') ?? null) as Project['userId'],
    latestSessionId: (readValue(record, 'latest_session_id') ?? readValue(record, 'latestSessionId') ?? null) as Project['latestSessionId'],
    lastMessageAt: (readValue(record, 'last_message_at') ?? readValue(record, 'lastMessageAt') ?? null) as Project['lastMessageAt'],
    createdAt: normalizeDate(createdAtSource),
    updatedAt: normalizeDate(updatedAtSource),
  };
};
